common/raw_hashtable.h

```diff
@@ -0,0 +1,1302 @@
+#ifndef NDEBUG
+  CARBON_DCHECK(debug_empty_count == (old_size - count));
+  CARBON_DCHECK(llvm::count(llvm::ArrayRef(new_metadata, new_size),
+                            MetadataGroup::Empty) ==
+                debug_empty_count +
+                    static_cast<ssize_t>(probed_indices.size()) + old_size);
+#endif
+
+  // If the keys or values are trivially relocatable, we do a bulk memcpy of
+  // them into place. This will copy them into both possible locations, which is
+  // fine. One will be empty and clobbered if reused or ignored. The other will
+  // be the one used. This might seem like it needs it to be valid for us to
+  // create two copies, but it doesn't. This produces the exact same storage as
+  // copying the storage into the wrong location first, and then again into the
+  // correct location. Only one is live and only one is destroyed.
+  if constexpr (EntryT::IsTriviallyRelocatable) {
+    memcpy(new_entries, old_entries, old_size * sizeof(EntryT));
+    memcpy(new_entries + old_size, old_entries, old_size * sizeof(EntryT));
+  }
+
+  // We have to use the normal insert for anything that was probed before, but
+  // we know we'll find an empty slot, so leverage that. We extract the probed
+  // keys from the bottom of the old keys storage.
+  for (ssize_t old_index : probed_indices) {
+    // We may end up needing to do a sequence of re-inserts, swapping out keys
+    // and values each time, so we enter a loop here and break out of it for the
+    // simple cases of re-inserting into a genuinely empty slot.
+    EntryT* new_entry = InsertIntoEmpty(old_entries[old_index].key());
+    new_entry->MoveFrom(std::move(old_entries[old_index]));
+  }
+  CARBON_DCHECK(count ==
+                (new_size - llvm::count(llvm::ArrayRef(new_metadata, new_size),
+                                        MetadataGroup::Empty)));
+  growth_budget_ -= count;
+  CARBON_DCHECK(growth_budget_ ==
+                (GrowthThresholdForAllocSize(new_size) -
+                 (new_size - llvm::count(llvm::ArrayRef(new_metadata, new_size),
+                                         MetadataGroup::Empty))));
+  CARBON_DCHECK(growth_budget_ > 0 &&
+                "Must still have a growth budget after rehash!");
+
+  if (!old_small) {
+    // Old isn't a small buffer, so we need to deallocate it.
+    Deallocate(old_storage, old_size);
+  }
+
+  // And lastly insert the lookup_key into an index in the newly grown map and
+  // return that index for use.
+  --growth_budget_;
+  return InsertIntoEmpty(lookup_key);
+}
+
+// Reset a table to its original state, including releasing any allocated
+// memory.
+template <typename InputBaseT, ssize_t SmallSize>
+auto TableImpl<InputBaseT, SmallSize>::ResetImpl() -> void {
+  this->Destroy();
+
+  // Re-initialize the whole thing.
+  CARBON_DCHECK(this->small_alloc_size() == SmallSize);
+  this->Construct(small_storage());
+}
+
+// We specialize the absence of any small size buffer to force to zero and null.
```

Comment:
I'm not sure what "force to zero and null" means.
```suggestion
// We define a specialization when there is no small size buffer.
```
It doesn't look like `Construct` looks at its argument when `small_alloc_size_` is 0. Is this to avoid calling `small_storage()` when there is no buffer? If so, say that.